### PR TITLE
Design for PrivateLink Connection Status History table

### PR DESCRIPTION
**Pull request by rjobanp**
### Motivation

  * This PR adds a known-desirable feature. See https://github.com/MaterializeInc/materialize/issues/19022 for context.

### Tips for reviewer

[Rendered Version](https://github.com/rjobanp/materialize/blob/privatelink-status-design-doc/doc/developer/design/20231103_privatelink_status_table.md)

Some code links for helpful context:
- `VpcEndpointStatus` definition: https://github.com/MaterializeInc/materialize/blob/main/src/cloud-resources/src/crd/vpc_endpoint.rs#L50
- Existing vpc_endpoint related methods on the `CloudResourceController`: https://github.com/MaterializeInc/materialize/blob/main/src/orchestrator-kubernetes/src/cloud_resource_controller.rs
- the `drain_statement_log` method used as a model for how to flush events to storage on an interval: https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/coord/statement_logging.rs#L122

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (benesch) [inline doc/developer/design/20231103_privatelink_status_table.md:51]: Do we have _some_ cleanup story for these storage-managed collections, though? It'd be good to expire the status after a certain number of days, so that the history doesn't grow forever without bound.

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (benesch) [inline doc/developer/design/20231103_privatelink_status_table.md:63]: You probably just want one or the other of these fields. I think the existing `CloudResourceController` public API is in terms of `id: GlobalId`, and so that's probably what you want for the events, too.

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (benesch) [inline doc/developer/design/20231103_privatelink_status_table.md:65]:
```suggestion
    time: DateTime<Utc>,
```

`time` is what `ClusterEvent` uses, so probably best to align with that.

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (benesch) [inline doc/developer/design/20231103_privatelink_status_table.md:107]: Yes, definitely, keeping the state changes around seems worthwhile!

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (benesch) [inline doc/developer/design/20231103_privatelink_status_table.md:123]: One other option: read in the table on startup and use it to initialize the in-memory current state for each VPC endpoint.

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (benesch) [inline doc/developer/design/20231103_privatelink_status_table.md:90]: I have a slightly different implementation to propose! I would use something like [`governor`](https://docs.rs/governor/latest/governor/struct.Quota.html) to rate limit the number of updates, with a bit of burst capacity. That way a few state transitions in quick succession as a VpcEndpoint becomes healthy are emitted immediately, but if a VpcEndpoint get stuck in some sort of hot fail loop with a lot of status changes the coordinator doesn't get overwhelmed.

I'd do the rate limiting on the transmitter side, so that the coordinator can blindly translate events on the stream to writes to the storage managed collection—e.g.:

```rust
struct VpcEndpointEvent {
    status: VpcEndpointStatus,
    time: DateTime<Utc>
}

Message::VpcEndpointEvents(BTreeMap<GlobalId, VpcEndpointEvent>)
```

You also probably want the `watch_vpc_endpoints` task to coalesce multiple events for the same connection ID in each rate-limited batch—I _think_? At least, there doesn't seem to be much need for presenting the full history of state transitions, rather than just the states that exist for long enough to make it through the rate limiter.

**Review by benesch — COMMENTED** [review]
Very nice!

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (jkosh44) [inline doc/developer/design/20231103_privatelink_status_table.md:70]: Can you add a sentence or two about how this map gets populated on startup/restart?

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (jkosh44) [reply doc/developer/design/20231103_privatelink_status_table.md:123]: +1

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (jkosh44) [reply doc/developer/design/20231103_privatelink_status_table.md:70]: Ah ok, I see the open question below. Whatever is decided, can you add a sentence or two here indicating whether we populate the map on startup/restart or not?

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (rjobanp) [reply doc/developer/design/20231103_privatelink_status_table.md:123]: That does seem like the best option! I just wasn't sure of the performance tradeoff of doing that on startup -- we'd effectively need to query the table for the most-recent row of each distinct VPC endpoint. Since it sounds like that is acceptable I'll adjust the design accordingly - thanks!

**Comment by rjobanp** [comment]
> We'll also want to verify that our VPC Endpoint controller peridiocally resyncs to update the CR status. From a super quick read through, I'm not sure if an accepted / running connection ever gets reconciled again, so we might be missing events.

@pH14 I also thought this, but from what I can tell the reconciliation loop should be run about once-per hour since that's the default behavior of our k8s-controller, even when the `apply` method returns `Ok(None)` and doesn't explicitly request a re-queuing:
https://github.com/MaterializeInc/k8s-controller/blob/689ddd0d7a63b892b531125598fdd3387cd212c4/src/controller.rs#L238-L248

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (rjobanp) [reply doc/developer/design/20231103_privatelink_status_table.md:51]: Yes good point I'll add a note about cleanup, it looks like I can leverage the Storage Controller's `partially_truncate_status_history` method used for doing the same to the source/sink status history tables: https://github.com/MaterializeInc/materialize/blob/9816c627c02e4f030725f65505abf34be9cdcf63/src/storage-controller/src/lib.rs#L2516

Comment (rjobanp) [reply doc/developer/design/20231103_privatelink_status_table.md:51]: though it keeps a certain number of status updates per unique entity rather than doing a time-based expiration

Comment (benesch) [reply doc/developer/design/20231103_privatelink_status_table.md:51]: > though it keeps a certain number of status updates per unique entity rather than doing a time-based expiration

Oh that seems fine! And worth aligning with the existing `status_history` tables.

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (benesch) [reply doc/developer/design/20231103_privatelink_status_table.md:123]: Should be fine as long as we truncate regularly!

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (rjobanp) [reply doc/developer/design/20231103_privatelink_status_table.md:90]: Great idea - reworking the design to include this

**Comment by rjobanp** [comment]
@benesch @jkosh44 Thanks for the feedback - I believe I've addressed your comments. One thing to note is that I 'moved' the in-memory map from the `CloudResourceController.watch_vpc_endpoints` to the new task in the Adapter Coordinator since it seems to have it initialize from the Storage Controller there.

---


> Curious if there's a way to link this up with https://github.com/MaterializeInc/materialize/issues/21617 / https://github.com/MaterializeInc/materialize/issues/21354, where the design here is the implementation detail of how PrivateLink connection history is fed into a more generic system collection. And maybe mz_aws_privatelink_connection_status_history is then a view on top of that data that prettifies PrivateLink connection history specifically. Food for thought, haven't personally thought about this problem long enough to have a strong opinion!

@pH14 A more generic system collection could make sense to link those up, however my initial reaction is that it diverges from the implementation of `mz_sink_status_history` and `mz_source_status_history` which currently exist as separate collections. Refactoring those to be views on top of the generic collection may not be trivial, so it could lead to some scope creep here.

**Review by jkosh44 — APPROVED** [review]
LGTM!

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (benesch) [inline doc/developer/design/20231103_privatelink_status_table.md:59]: You'll want to make sure the `watch_vpc_endpoints` method has a return type like

```sql
BoxStream<'static, Result<VpcEndpointEvent, anyhow::Error>>
```

where the `'static` will let you move the stream into a task. Otherwise the stream will keep a reference to `'Self` (the `&CloudResourcesController`), and you won't be able to move thestream into the task.

**Review by benesch — APPROVED** [review]
Very nice! Make sure you get storage eyes (@guswynn, @sploiselle, @bkirwi have all touched the source/sink status code in addition to @petrosagg) on this before getting too deep, but this looks really solid.

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (rjobanp) [reply doc/developer/design/20231103_privatelink_status_table.md:59]: awesome thanks for the tip!

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (guswynn) [inline doc/developer/design/20231103_privatelink_status_table.md:57]: Will it only be truncated on startup, like the other status history tables? (this is fine, just wondering)

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (guswynn) [inline doc/developer/design/20231103_privatelink_status_table.md:96]: this all seems good to me! couple of things to be careful with during impl:
- the task can't allow you return an error, it must be permanently running
- it will need some kind of shared access to the catalog so it can begin and start collecting for connections as they are created and dropped
- it will probably be nice to use https://docs.rs/tokio-stream/latest/tokio_stream/struct.StreamMap.html in the impl of this task, message me if you want to discuss in detail!
- i don't think we need to bring the `govenor` crate, we can just have a min interval that we will emit new events to be written to the coordinator, and if multiple events come for the same connection in that interval, we will just choose the last one. i think it would be hard to safely wrap a kube watch with govenor, without possibly causing weird behavior like network timeouts

it would also be nice if this doc clarified _why_ we need to initialize this task with the last event (i assume its entirely to avoid re-emitting the last event?). I assume this will just have the coordinator reach into the storage controller it holds (I think the order of bootstrapping makes this not too bad)

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (guswynn) [inline doc/developer/design/20231103_privatelink_status_table.md:111]: a separate task that polls `list_vpc_endpoints` is still preferable to this; if watches are annoying thats what we should do

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (guswynn) [reply doc/developer/design/20231103_privatelink_status_table.md:96]: for my last question i see now its a resolved open question!

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (guswynn) [inline doc/developer/design/20231103_privatelink_status_table.md:141]: perhaps the controller for this checks them all at the same time?

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (guswynn) [inline doc/developer/design/20231103_privatelink_status_table.md:150]: again, i think dropping events and just choosing the most recent one is the best route here. Its much simpler, and I have horror stories about delaying polling futures/streams causing errors deep in the stack (like server-side timeouts)

**Review by guswynn — APPROVED** [review]
I don't think we need govenor, but other than that this looks good!

please message me if you have any questions about the implementation!

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (benesch) [reply doc/developer/design/20231103_privatelink_status_table.md:96]: > it will need some kind of shared access to the catalog so it can begin and start collecting for connections as they are created and dropped

Nope, it shouldn't! Should be possible to configure the Kubernetes watch to listen for all `VPCEndpoint` events in the namespace, _including_ those that are create or deletion events. Here's the analogous code in the `KubernetesOrchestrator`:

https://github.com/MaterializeInc/materialize/blob/a47c0a852d8c851c54ef4786dd630bc26b796d7e/src/orchestrator-kubernetes/src/lib.rs#L1476-L1488

> it will probably be nice to use https://docs.rs/tokio-stream/latest/tokio_stream/struct.StreamMap.html in the impl of this task, message me if you want to discuss in detail!

(There should only be the _one_ watch, so you shouldn't need to be managing multiple streams.)



> * i don't think we need to bring the `govenor` crate, we can just have a min interval that we will emit new events to be written to the coordinator, and if multiple events come for the same connection in that interval, we will just choose the last one. i think it would be hard to safely wrap a kube watch with govenor, without possibly causing weird behavior like network timeouts

Discussed here: https://github.com/MaterializeInc/materialize/pull/22936#discussion_r1382288410

The reason I proposed using something like `governor` is because I want us to build the muscle of reacting to things in real time. It seems like a shame to build a real time system that has a (say) forced poll interval of 5s. Since the underlying Kubernetes tech here supports proper events, it'd be great to be able to usually bring those events in in real time, but with some protections to prevent overload.

> i think it would be hard to safely wrap a kube watch with govenor, without possibly causing weird behavior like network timeouts

I think it should be fine? Here's the code I had in mind:

```rust
let (coord_tx, coord_rx) = mpsc::channel();
tokio::spawn(async {
    let watch = cloud_resource_controller.watch_vpc_endpoints();
    let rate_limiter = governor::RateLimiter::new(...);
    loop {
        // Wait for an event to become available.
        let event = watch.next().await;
        
        // Wait until we're permitted to tell the coordinator about the event.
        rate_limiter.until_ready().await;
       
        // Drain any additional events from the queue that accumulated while
        // waiting for the rate limiter, deduplicating by VPC endpoint ID.
        let mut events = BTreeMap::new();
        events.insert(event.id, event);
        while let Some(event) = watch.next().now_or_never(); {
            events.insert(event.id, event);
        }

        // Send the event batch to the coordinator.
        coord_tx.send(Message::VpcEndpointEvents(events));
    }
})
```

`StreamExt::next` is cancel safe so no issue there. `watch.next().await` will poll to completion so shouldn't be any issues with network timeouts there. Kubernetes watches are also just long running HTTP requests, so there's no need to poll them on any frequency to keep them alive—and in fact the Kubernetes API server will force terminate them occasionally, to ensure that all clients have good retry behavior.

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (rjobanp) [reply doc/developer/design/20231103_privatelink_status_table.md:57]: Yep just on startup is the plan, like the other status history tables

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (rjobanp) [reply doc/developer/design/20231103_privatelink_status_table.md:96]: I think @benesch responded to most of the concerns you raised @guswynn, though I'll reword the doc to make it clearer that the kubernetes watch will be on all VPCEndpoints in the namespace, including newly created ones, so I don't think we need to have access to the catalog and don't need to use `StreamMap`.

The `watcher` method in kube-runtime https://docs.rs/kube/latest/kube/runtime/fn.watcher.html also does provide some built-in error handling for us, so retries and connection issues should be handled before we receive events on the stream. 

Re: `governor` - thanks for the code sample @benesch -- I think that approach looks reasonable and if we hit issues with the rate-limiting logic during implementation we can re-evaluate.

Comment (guswynn) [reply doc/developer/design/20231103_privatelink_status_table.md:96]: ah sounds good!

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (rjobanp) [reply doc/developer/design/20231103_privatelink_status_table.md:141]: Potentially, though that seems a little unlikely since they should each have had their own reconciliation happen when they were created? @jubrad any ideas? You can run
```
kubectl get vpcendpoints -A -o custom-columns=":metadata.name,:status.conditions"
```
to replicate

Comment (jubrad) [reply doc/developer/design/20231103_privatelink_status_table.md:142]: I'm seeing this: `lastTransitionTime:2023-09-19T00:16:54Z `... this is odd. I actually wouldn't be that surprised if everything ran through reconciliation at pretty much the same time when the environment controller came up after this code was deployed. It's also possible no privatelink connections have [been created since](https://grafana.dev.materialize.com/d/d6aadd32-0832-43d7-8bb1-63ec7cf19d05/privatelink-health?orgId=1&from=now-30d&to=now). However, I would expect that to have occurred right after the code had merged, which was aug 23rd not sept 19th.

Here's where this should be set.
https://github.com/MaterializeInc/cloud/pull/7174/files#diff-2bdf70ebf1d6a05b7c26a2fecdb1f31ded5cc5c683a3e8bb036ddcb13b9a772bR394

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (rjobanp) [reply doc/developer/design/20231103_privatelink_status_table.md:112]: resolving based on discussion in https://github.com/MaterializeInc/materialize/pull/22936#discussion_r1385786183 (and assuming that k8s watch will work :) )

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (rjobanp) [reply doc/developer/design/20231103_privatelink_status_table.md:151]: resolving based on discussion in https://github.com/MaterializeInc/materialize/pull/22936#discussion_r1385786183

File: doc/developer/design/20231103_privatelink_status_table.md
Comment (rjobanp) [reply doc/developer/design/20231103_privatelink_status_table.md:142]: Yep that's what I see too. I'm not going to block merging this design doc on this, but will see if I can add validation for the `last_transition_time` field in `test_privatelink` in the cloud repo to make sure it's being set. Hopefully it's just something about the deployment time that caused all the us-east-1 objects to get the same timestamp

**Pull request merged**
